src/Runner.Worker/ExpressionFunctions/HashFiles.cs

```diff
@@ -0,0 +1,139 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Minimatch;
+using System.IO;
+using System.Security.Cryptography;
+using GitHub.DistributedTask.Expressions2.Sdk;
+using GitHub.DistributedTask.Pipelines.ContextData;
+using GitHub.DistributedTask.Pipelines.ObjectTemplating;
+using GitHub.Runner.Common;
+using GitHub.Runner.Sdk;
+using System.Reflection;
+using System.Threading;
+using System.Runtime.Serialization;
+
+namespace GitHub.Runner.Worker.Handlers
+{
+    public class FunctionTrace : ITraceWriter
+    {
+        private GitHub.DistributedTask.Expressions2.ITraceWriter _trace;
+
+        public FunctionTrace(GitHub.DistributedTask.Expressions2.ITraceWriter trace)
+        {
+            _trace = trace;
+        }
+        public void Info(string message)
+        {
+            _trace.Info(message);
+        }
+
+        public void Verbose(string message)
+        {
+            _trace.Info(message);
+        }
+    }
+
+    public sealed class HashFiles : Function
+    {
+        protected sealed override Object EvaluateCore(
+            EvaluationContext context,
+            out ResultMemory resultMemory)
+        {
+            resultMemory = null;
+
+            // hashFiles() only works on the runner and only works with files under GITHUB_WORKSPACE
+            // Since GITHUB_WORKSPACE is set by runner, I am using that as the fact of this code runs on server or runner.
+            if (context.State is DistributedTask.ObjectTemplating.TemplateContext templateContext &&
+                templateContext.ExpressionValues.TryGetValue(PipelineTemplateConstants.GitHub, out var githubContextData) &&
+                githubContextData is DictionaryContextData githubContext &&
+                githubContext.TryGetValue(PipelineTemplateConstants.Workspace, out var workspace) == true &&
+                workspace is StringContextData workspaceData)
+            {
+                string searchRoot = workspaceData.Value;
+                bool followSymlink = false;
+                string pattern = "";
+                if (Parameters.Count == 1)
+                {
+                    pattern = Parameters[0].Evaluate(context).ConvertToString();
+                }
+                else
+                {
+                    if (string.Equals(Parameters[0].Evaluate(context).ConvertToString(), "--followSymbolicLinks", StringComparison.OrdinalIgnoreCase))
```

Comment:
if it's not `--follow-symbolic-links` then is it just silently ignored?

Reply:
should be `--follow-symbolic-links` not `--followSymbolicLinks` correct?

Reply:
Oops.. i must copied that string from somewhere...